### PR TITLE
fix(chat): inject special tokens (bos_token, eos_token) into chat template context

### PR DESCRIPTION
**Pull request by CatherineSue**
### Problem

Chat templates that reference `{{ bos_token }}` or `{{ eos_token }}` (e.g., Llama 4, Llama 2/3) rendered these as empty strings because SMG did not pass special tokens into the Jinja template context.

HuggingFace Python does this automatically:
```python
template_kwargs = {**self.special_tokens_map, **kwargs}
```

This caused two issues:

1. **Missing BOS token**: 1-token-per-request difference between SMG gRPC and vLLM HTTP for Llama 4 (928,592 vs 929,492 input tokens on MMMU val).

2. **Mistral tool calling failure**: With BOS correctly injected, Mistral-7B-Instruct-v0.3 outputs `[TOOL_CALLS]` prefix (correct model behavior), which conflicts with guided decoding JSON schema constraint for `tool_choice=required`. Without BOS, the model skipped `[TOOL_CALLS]` and went straight to JSON — accidental success that masked the missing BOS bug. Note: sglang also skips `tool_choice=required` tests for Mistral due to the same whitespace issue.

### Solution

- Add `special_tokens` field to `ChatTemplateParams` and inject `bos_token`, `eos_token`, `unk_token`, `pad_token` into the minijinja context. Missing tokens use `Value::UNDEFINED` so `{% if bos_token is defined %}` guards work correctly.
- Add Mistral parallel tool chat template and configure e2e tests to use it with `--chat-template`.
- Add `--structured-outputs-config` for vLLM and `--constrained-json-disable-any-whitespace` for sglang in Mistral model spec.

Supersedes #818

## Changes

- `crates/tokenizer/src/chat_template.rs` — Add `special_tokens` to `ChatTemplateParams`, inject into context. 3 unit tests.
- `model_gateway/src/routers/grpc/utils/chat_utils.rs` — Pass `tokenizer.get_special_tokens()` for chat completions
- `model_gateway/src/routers/grpc/utils/message_utils.rs` — Same for Messages API
- `bindings/golang/src/tokenizer.rs` — Same for Go bindings
- `crates/tokenizer/tests/chat_template_integration.rs` — 2 integration tests for BOS injection
- `e2e_test/fixtures/chat_templates/tool_chat_template_mistral_parallel.jinja` — Mistral parallel tool template
- `e2e_test/chat_completions/test_function_calling.py` — Configure Mistral test to use chat template
- `e2e_test/infra/model_specs.py` — Add structured output config for Mistral (vLLM + sglang)

## Test Plan

```bash
cargo test -p llm-tokenizer --lib -- "chat_template::tests::test_special_tokens"
cargo test -p llm-tokenizer --test chat_template_integration -- "test_bos_token"
```

<details>
<summary>Manual verification: Mistral tool_choice=required</summary>

```bash
vllm serve /raid/models/mistralai/Mistral-7B-Instruct-v0.3 --port 8080 --grpc \
  --structured-outputs-config '{"disable_any_whitespace": true, "backend": "xgrammar"}'

cargo run --bin smg -- --host 0.0.0.0 --port 3002 \
  --worker-urls grpc://127.0.0.1:8080 \
  --model-path /raid/models/mistralai/Mistral-7B-Instruct-v0.3/ \
  --chat-template e2e_test/fixtures/chat_templates/tool_chat_template_mistral_parallel.jinja

curl -s http://localhost:3002/v1/chat/completions -H 'Content-Type: application/json' -d '{
  "model": "/raid/models/mistralai/Mistral-7B-Instruct-v0.3",
  "messages": [{"role": "user", "content": "What is the weather in Tokyo?"}],
  "tools": [{"type":"function","function":{"name":"get_weather","description":"Get weather","parameters":{"type":"object","properties":{"location":{"type":"string"}},"required":["location"]}}}],
  "tool_choice": "required", "max_tokens": 200
}' | jq
```

Response:
```json
{"choices": [{"message": {"role": "assistant", "tool_calls": [{"id": "call_...", "type": "function", "function": {"name": "get_tourist_attractions", "arguments": "{\"location\":\"Tokyo\",\"limit\":5}"}}]}, "finish_reason": "tool_calls"}]}
```

</details>

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes

</details>

## Summary by CodeRabbit

* **New Features**
  * Chat templates now receive and expose special tokens (BOS, EOS, PAD, UNK) during rendering; tokenizer token discovery now prefers values from tokenizer config when present.
  * Added a Mistral-parallel chat template for multi-turn conversations, tool calls, and tool results.

* **Tests**
  * Added integration and unit tests validating special-token injection and defined/undefined behavior.

* **Chores**
  * Updated Mistral test runtime and model runtime settings used by end-to-end tests.

### CODE REVIEW

**Comment by hyeongyun0916** [comment]
The current implementation injects only 4 tokens (bos_token, eos_token, unk_token, pad_token). Is it okay to leave out (sep_token, cls_token, mask_token...) ?

File: crates/tokenizer/src/huggingface.rs [hunk 1]
```diff
@@ -171,22 +178,69 @@ impl HuggingFaceTokenizer {
...
+                .1
+                .or_else(|| find_token(&["</s>", "<|endoftext|>", "<EOS>", "[SEP]"])),
+            unk_token: config_tokens
+                .2
+                .or_else(|| find_token(&["<unk>", "<UNK>", "[UNK]"])),
             sep_token: find_token(&["[SEP]", "<sep>", "<SEP>"]),
-            pad_token: find_token(&["<pad>", "<PAD>", "[PAD]"]),
+            pad_token: config_tokens
+                .3
+                .or_else(|| find_token(&["<pad>", "<PAD>", "[PAD]"])),
```
[the remark this answers is not in the document]
Comment (CatherineSue) [reply crates/tokenizer/src/huggingface.rs:204]: Too trivial, unlike to happen

**Comment by CatherineSue** [comment]
> The current implementation injects only 4 tokens (bos_token, eos_token, unk_token, pad_token). Is it okay to leave out (sep_token, cls_token, mask_token...) ?

I don't see a lot of references to those tokens. We will add them when used. It is not worth to add a lot of abstraction for the special tokens.
